File: ios/Sources/RisemaxiGraft/GraftPlugin.swift
import Foundation
import Capacitor

@objc(GraftPlugin)
public class GraftPlugin: CAPPlugin, CAPBridgedPlugin {
    public static let tag = "Graft"
    public static let userDefaultsPrefix = "RisemaxiGraft" // DO NOT CHANGE

    public let identifier = "GraftPlugin"
    public let jsName = "Graft"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "clearBlockedBundles", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteBundle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "downloadBundle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBlockedBundles", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getChannel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentBundle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDownloadedBundles", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getInstallId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getNextBundle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVersionCode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVersionName", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isSyncing", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "ready", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reset", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setChannel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setNextBundle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sync", returnType: CAPPluginReturnPromise)
    ]

    private let eventDownloadBundleProgess = "downloadBundleProgress"
    private let eventNextBundleSet = "nextBundleSet"
    private let eventReloaded = "reloaded"

    private var implementation: Graft?

    override public func load() {
        let implementation = Graft(config: graftConfig(), plugin: self)
        self.implementation = implementation

        implementation.handleLoad()

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleAppWillEnterForeground),
            name: UIApplication.willEnterForegroundNotification,
            object: nil
        )
    }

    @objc func clearBlockedBundles(_ call: CAPPluginCall) {
        guard let implementation = graft(call) else {
            return
        }
        implementation.clearBlockedBundles()
        call.resolve()
    }

    @objc func deleteBundle(_ call: CAPPluginCall) {
        guard let bundleId = call.getString("bundleId") else {
            call.reject(CustomError.bundleIdMissing.localizedDescription)
            return
        }
        graft(call)?.deleteBundle(DeleteBundleOptions(bundleId: bundleId), completion: emptyCompletion(call))
    }

    @objc func downloadBundle(_ call: CAPPluginCall) {
        guard let implementation = graft(call) else {
            return
        }
        guard let bundleId = call.getString("bundleId") else {
            call.reject(CustomError.bundleIdMissing.localizedDescription)
            return
        }
        guard let checksum = call.getString("checksum") else {
            call.reject(CustomError.checksumMissing.localizedDescription)
            return
        }
        guard let url = call.getString("url") else {
            call.reject(CustomError.urlMissing.localizedDescription)
            return
        }
        Task {
            do {
                try await implementation.downloadBundle(DownloadBundleOptions(bundleId: bundleId, checksum: checksum, url: url))
                call.resolve()
            } catch {
                rejectCall(call, error)
            }
        }
    }

    @objc func getBlockedBundles(_ call: CAPPluginCall) {
        graft(call)?.getBlockedBundles(completion: resultCompletion(call))
    }

    @objc func getChannel(_ call: CAPPluginCall) {
        graft(call)?.getChannel(completion: resultCompletion(call))
    }

    @objc func getCurrentBundle(_ call: CAPPluginCall) {
        graft(call)?.getCurrentBundle(completion: resultCompletion(call))
    }

    @objc func getDownloadedBundles(_ call: CAPPluginCall) {
        graft(call)?.getDownloadedBundles(completion: resultCompletion(call))
    }

    @objc func getInstallId(_ call: CAPPluginCall) {
        graft(call)?.getInstallId(completion: resultCompletion(call))
    }

    @objc func getNextBundle(_ call: CAPPluginCall) {
        graft(call)?.getNextBundle(completion: resultCompletion(call))
    }

    @objc func getVersionCode(_ call: CAPPluginCall) {
        graft(call)?.getVersionCode(completion: resultCompletion(call))
    }

    @objc func getVersionName(_ call: CAPPluginCall) {
        graft(call)?.getVersionName(completion: resultCompletion(call))
    }

    @objc func isSyncing(_ call: CAPPluginCall) {
        graft(call)?.isSyncing(completion: resultCompletion(call))
    }

    @objc func ready(_ call: CAPPluginCall) {
        graft(call)?.ready(completion: resultCompletion(call))
    }

    @objc func reload(_ call: CAPPluginCall) {
        guard let implementation = graft(call) else {
            return
        }
        implementation.reload()
        call.resolve()
    }

    @objc func reset(_ call: CAPPluginCall) {
        guard let implementation = graft(call) else {
            return
        }
        implementation.reset()
        call.resolve()
    }

    @objc func setChannel(_ call: CAPPluginCall) {
        graft(call)?.setChannel(SetChannelOptions(channel: call.getString("channel")), completion: emptyCompletion(call))
    }

    @objc func setNextBundle(_ call: CAPPluginCall) {
        graft(call)?.setNextBundle(SetNextBundleOptions(call), completion: emptyCompletion(call))
    }

    @objc func sync(_ call: CAPPluginCall) {
        guard let implementation = graft(call) else {
            return
        }
        Task {
            do {
                resolveCall(call, try await implementation.sync(SyncOptions(call)))
            } catch {
                rejectCall(call, error)
            }
        }
    }

    func notifyDownloadBundleProgressListeners(_ event: DownloadBundleProgressEvent) {
        if let event = event.toJSObject() as? JSObject {
            notifyListeners(eventDownloadBundleProgess, data: event, retainUntilConsumed: false)
        }
    }

    func notifyNextBundleSetListeners(_ event: NextBundleSetEvent) {
        notifyListeners(eventNextBundleSet, data: event.toJSObject(), retainUntilConsumed: false)
    }

    func notifyReloadedListeners() {
        notifyListeners(eventReloaded, data: JSObject(), retainUntilConsumed: true)
    }

    @objc private func handleAppWillEnterForeground() {
        implementation?.handleAppWillEnterForeground()
    }

    private func graftConfig() -> GraftConfig {
        var config = GraftConfig()

        config.autoBlockRolledBackBundles = getConfig().getBoolean("autoBlockRolledBackBundles", config.autoBlockRolledBackBundles)
        config.autoDeleteBundles = getConfig().getBoolean("autoDeleteBundles", config.autoDeleteBundles)
        config.autoUpdateStrategy = getConfig().getString("autoUpdateStrategy", config.autoUpdateStrategy) ?? config.autoUpdateStrategy
        config.defaultChannel = getConfig().getString("defaultChannel", config.defaultChannel)
        config.httpTimeout = getConfig().getInt("httpTimeout", config.httpTimeout)
        config.publicKey = getConfig().getString("publicKey", config.publicKey)
        config.readyTimeout = getConfig().getInt("readyTimeout", config.readyTimeout)
        config.serverUrl = getConfig().getString("serverUrl", config.serverUrl)

        return config
    }

    private func graft(_ call: CAPPluginCall) -> Graft? {
        if implementation == nil {
            call.reject(CustomError.notInitialized.localizedDescription)
        }
        return implementation
    }

    private func emptyCompletion(_ call: CAPPluginCall) -> (Error?) -> Void {
        return { error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            call.resolve()
        }
    }

    private func resultCompletion(_ call: CAPPluginCall) -> (Result?, Error?) -> Void {
        return { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            guard let result = result else {
                call.resolve()
                return
            }
            self.resolveCall(call, result)
        }
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result) {
        guard let value = result.toJSObject() as? JSObject else {
            call.resolve()
            return
        }
        call.resolve(value)
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", GraftPlugin.tag, "] ", error)
        var message = error.localizedDescription
        if let urlError = error as? URLError, urlError.code == .timedOut {
            message = CustomError.httpTimeout.localizedDescription
        }
        call.reject(message)
    }
}

File: android/src/main/java/com/risemaxi/graft/GraftPlugin.java
package com.risemaxi.graft;

import android.webkit.WebView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.webkit.WebViewCompat;
import androidx.webkit.WebViewFeature;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.WebViewListener;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.risemaxi.graft.classes.events.DownloadBundleProgressEvent;
import com.risemaxi.graft.classes.events.NextBundleSetEvent;
import com.risemaxi.graft.classes.options.DeleteBundleOptions;
import com.risemaxi.graft.classes.options.DownloadBundleOptions;
import com.risemaxi.graft.classes.options.SetChannelOptions;
import com.risemaxi.graft.classes.options.SetNextBundleOptions;
import com.risemaxi.graft.classes.options.SyncOptions;
import com.risemaxi.graft.interfaces.EmptyCallback;
import com.risemaxi.graft.interfaces.NonEmptyCallback;
import com.risemaxi.graft.interfaces.Result;
import java.net.SocketTimeoutException;
import java.util.Collections;

@CapacitorPlugin(name = "Graft")
public class GraftPlugin extends Plugin {

    public static final String TAG = "Graft";
    public static final String SHARED_PREFERENCES_NAME = "RisemaxiGraft"; // DO NOT CHANGE
    public static final String RELEASE_IDENTITY_KEY = "__graft__"; // must match src/identity.ts

    public static final String ERROR_BUNDLE_EXISTS = "bundle already exists.";
    public static final String ERROR_BUNDLE_ID_MISSING = "bundleId must be provided.";
    public static final String ERROR_BUNDLE_INDEX_HTML_MISSING = "The bundle does not contain an index.html file.";
    public static final String ERROR_BUNDLE_NOT_FOUND = "bundle not found.";
    public static final String ERROR_CHANNEL_MISSING = "channel must be configured.";
    public static final String ERROR_CHECKSUM_MISSING = "checksum must be provided.";
    public static final String ERROR_CHECKSUM_MISMATCH = "Checksum mismatch.";
    public static final String ERROR_DOWNLOAD_FAILED = "Bundle could not be downloaded.";
    public static final String ERROR_HTTP_TIMEOUT = "Request timed out.";
    public static final String ERROR_INSTALL_FAILED = "Bundle could not be installed.";
    public static final String ERROR_MANIFEST_EXPIRED = "The manifest is not valid at the current time.";
    public static final String ERROR_MANIFEST_MISMATCH = "The manifest does not describe an acceptable release.";
    public static final String ERROR_MANIFEST_NAME_COLLISION =
        "A file in the release has the same name as the manifest, so one overwrites the other. Serve the manifest as graft-manifest.json.";
    public static final String ERROR_MANIFEST_URL_INVALID = "The manifest URL is not on the configured server.";
    public static final String ERROR_NOT_INITIALIZED = "Graft failed to initialize.";
    public static final String ERROR_PATCH_CHECKSUM_MISMATCH = "A patched file did not match the manifest.";
    public static final String ERROR_PATCH_FAILED = "The patch could not be applied.";
    public static final String ERROR_PUBLIC_KEY_INVALID = "Invalid public key.";
    public static final String ERROR_PUBLIC_KEY_MISSING = "publicKey must be configured.";
    public static final String ERROR_SERVER_URL_INVALID = "Invalid serverUrl.";
    public static final String ERROR_SERVER_URL_MISSING = "serverUrl must be configured.";
    public static final String ERROR_SIGNATURE_VERIFICATION_FAILED = "Signature verification failed.";
    public static final String ERROR_SYNC_IN_PROGRESS = "Sync is already in progress.";
    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String ERROR_URL_MISSING = "url must be provided.";

    public static final String EVENT_DOWNLOAD_BUNDLE_PROGRESS = "downloadBundleProgress";
    public static final String EVENT_NEXT_BUNDLE_SET = "nextBundleSet";
    public static final String EVENT_RELOADED = "reloaded";

    @Nullable
    private Graft implementation;

    private boolean webViewListenerRegistered = false;

    public void load() {
        try {
            implementation = new Graft(getGraftConfig(), this);
            publishReleaseIdentity(implementation);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    /**
     * Publishes the running bundle's identity to the page before any of its code runs, so an app can
     * read it synchronously instead of compiling a release identifier into its own bundle.
     */
    private void publishReleaseIdentity(@NonNull Graft graft) {
        if (!WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            Logger.warn(TAG, "This WebView cannot run a document-start script, so releaseIdentity() will return null.");
            return;
        }
        try {
            String script = "globalThis." + RELEASE_IDENTITY_KEY + " = " + graft.getReleaseIdentity() + ";";
            WebViewCompat.addDocumentStartJavaScript(getBridge().getWebView(), script, Collections.singleton(getBridge().getLocalUrl()));
        } catch (Exception exception) {
            Logger.error(TAG, "Failed to publish the release identity to the page.", exception);
        }
    }

    @Override
    protected void handleOnResume() {
        super.handleOnResume();
        try {
            if (implementation != null) {
                implementation.handleOnResume();
            }
            // Important: For some reason, the listener CANNOT be registered in the load() method
            // or constructor, it MUST be done here in onResume().
            if (!webViewListenerRegistered) {
                webViewListenerRegistered = true;
                getBridge().addWebViewListener(
                    new WebViewListener() {
                        @Override
                        public void onPageLoaded(WebView webView) {
                            if (implementation != null) {
                                implementation.handleOnPageLoaded();
                            }
                        }
                    }
                );
            }
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void clearBlockedBundles(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.clearBlockedBundles();
            call.resolve();
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void deleteBundle(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            String bundleId = call.getString("bundleId");
            if (bundleId == null) {
                call.reject(ERROR_BUNDLE_ID_MISSING);
                return;
            }
            graft.deleteBundle(new DeleteBundleOptions(bundleId), emptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void downloadBundle(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            String bundleId = call.getString("bundleId");
            if (bundleId == null) {
                call.reject(ERROR_BUNDLE_ID_MISSING);
                return;
            }
            String checksum = call.getString("checksum");
            if (checksum == null) {
                call.reject(ERROR_CHECKSUM_MISSING);
                return;
            }
            String url = call.getString("url");
            if (url == null) {
                call.reject(ERROR_URL_MISSING);
                return;
            }
            graft.downloadBundle(new DownloadBundleOptions(bundleId, checksum, url), emptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getBlockedBundles(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getBlockedBundles(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getChannel(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getChannel(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getCurrentBundle(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getCurrentBundle(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getDownloadedBundles(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getDownloadedBundles(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getInstallId(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getInstallId(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getNextBundle(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getNextBundle(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getVersionCode(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getVersionCode(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getVersionName(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.getVersionName(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void isSyncing(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.isSyncing(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void ready(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.ready(resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void reload(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.reload();
            call.resolve();
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void reset(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.reset();
            call.resolve();
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setChannel(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.setChannel(new SetChannelOptions(call.getString("channel")), emptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setNextBundle(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.setNextBundle(new SetNextBundleOptions(call), emptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void sync(PluginCall call) {
        Graft graft = requireImplementation(call);
        if (graft == null) {
            return;
        }
        try {
            graft.sync(new SyncOptions(call), resultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    public void notifyDownloadBundleProgressListeners(@NonNull DownloadBundleProgressEvent event) {
        notifyListeners(EVENT_DOWNLOAD_BUNDLE_PROGRESS, event.toJSObject(), false);
    }

    public void notifyNextBundleSetListeners(@NonNull NextBundleSetEvent event) {
        notifyListeners(EVENT_NEXT_BUNDLE_SET, event.toJSObject(), false);
    }

    public void notifyReloadedListeners() {
        notifyListeners(EVENT_RELOADED, new JSObject(), true);
    }

    @Nullable
    private Graft requireImplementation(@NonNull PluginCall call) {
        if (implementation == null) {
            call.reject(ERROR_NOT_INITIALIZED);
        }
        return implementation;
    }

    @NonNull
    private EmptyCallback emptyCallback(@NonNull PluginCall call) {
        return new EmptyCallback() {
            @Override
            public void success() {
                call.resolve();
            }

            @Override
            public void error(Exception exception) {
                rejectCall(call, exception);
            }
        };
    }

    @NonNull
    private <T extends Result> NonEmptyCallback<T> resultCallback(@NonNull PluginCall call) {
        return new NonEmptyCallback<>() {
            @Override
            public void success(@NonNull T result) {
                call.resolve(result.toJSObject());
            }

            @Override
            public void error(Exception exception) {
                rejectCall(call, exception);
            }
        };
    }

    @NonNull
    private GraftConfig getGraftConfig() {
        GraftConfig config = new GraftConfig();

        config.setAutoBlockRolledBackBundles(getConfig().getBoolean("autoBlockRolledBackBundles", config.getAutoBlockRolledBackBundles()));
        config.setAutoDeleteBundles(getConfig().getBoolean("autoDeleteBundles", config.getAutoDeleteBundles()));
        config.setAutoUpdateStrategy(getConfig().getString("autoUpdateStrategy", config.getAutoUpdateStrategy()));
        config.setDefaultChannel(getConfig().getString("defaultChannel", config.getDefaultChannel()));
        config.setHttpTimeout(getConfig().getInt("httpTimeout", config.getHttpTimeout()));
        config.setPublicKey(getConfig().getString("publicKey", config.getPublicKey()));
        config.setReadyTimeout(getConfig().getInt("readyTimeout", config.getReadyTimeout()));
        config.setServerUrl(getConfig().getString("serverUrl", config.getServerUrl()));

        return config;
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (exception instanceof SocketTimeoutException) {
            message = ERROR_HTTP_TIMEOUT;
        } else if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }
}
